source/MaterialXTest/GenReference.cpp

```diff
@@ -0,0 +1,134 @@
+//
+// TM & (c) 2017 Lucasfilm Entertainment Company Ltd. and Lucasfilm Ltd.
+// All rights reserved.  See LICENSE.txt for license.
+//
+
+#include <MaterialXTest/Catch/catch.hpp>
+
+#include <MaterialXTest/GenShaderUtil.h>
+
+#include <MaterialXCore/Document.h>
+
+#include <MaterialXFormat/File.h>
+
+#include <MaterialXGenShader/Util.h>
+#include <MaterialXGenShader/Shader.h>
+#include <MaterialXGenShader/ShaderStage.h>
+#include <MaterialXGenOsl/OslShaderGenerator.h>
+#include <MaterialXGenOsl/OslSyntax.h>
+
+namespace mx = MaterialX;
+
+const mx::ShaderPort* getShaderPort(const mx::ShaderStage& stage, const std::string& name)
+{
+    for (const auto& it : stage.getUniformBlocks())
+    {
+        const mx::VariableBlock& block = *it.second;
+        const mx::ShaderPort* port = block.find(name);
+        if (port)
+        {
+            return port;
+        }
+    }
+    for (const auto& it : stage.getInputBlocks())
+    {
+        const mx::VariableBlock& block = *it.second;
+        const mx::ShaderPort* port = block.find(name);
+        if (port)
+        {
+            return port;
+        }
+    }
+    for (const auto& it : stage.getOutputBlocks())
+    {
+        const mx::VariableBlock& block = *it.second;
+        const mx::ShaderPort* port = block.find(name);
+        if (port)
+        {
+            return port;
+        }
+    }
+    return nullptr;
+}
+
+TEST_CASE("GenShader: OSL Reference", "[genshader]")
```

Comment:
Unittest performing the generation of OSL reference library. Longer term we could move this into a build step instead when building the MaterialXGenOsl project.